syntax = "proto3";

package usermanagement;

service UserService {
    // Metodi per la gestione degli utenti
    rpc RegisterUser (UserRegisterRequest) returns (UserResponse);
    rpc UpdateUser (UserUpdateRequest) returns (UserResponse);
    rpc DeleteUser (UserIdentifier) returns (UserResponse);
    rpc LoginUser (UserLoginRequest) returns (UserResponse);
    
    // Metodi per il recupero dei dati
    rpc GetLatestValue (UserIdentifier) returns (StockValueResponse);
    rpc GetAverageValue (AverageRequest) returns (AverageResponse);
}


// Messaggi per le richieste
message UserRegisterRequest {
    string email = 1;
    string password = 2;
    string ticker = 3;
}

message UserUpdateRequest {
    string email = 1;
    string new_ticker = 2;
}

message UserLoginRequest{
    string email = 1;
    string password = 2;
}

message UserIdentifier {
    string email = 1;
}

message AverageRequest {
    string email = 1;
    int32 num_values = 2;
}


// Messaggi per le risposte
message UserResponse {
    bool success = 1;
    string message = 2;
}

message StockValueResponse {
    bool success = 1;
    string message = 2;
    string ticker = 3;
    float value = 4;
    string timestamp = 5;
}

message AverageResponse {
    bool success = 1;
    string message = 2;
    string ticker = 3;
    float average = 4;
}
